builder/api/src/main/java/io/helidon/builder/api/Prototype.java

```diff
@@ -0,0 +1,414 @@
+/*
+ * Copyright (c) 2023 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.builder.api;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.List;
+
+import io.helidon.common.HelidonServiceLoader;
+import io.helidon.common.config.Config;
+import io.helidon.common.config.ConfiguredProvider;
+import io.helidon.common.config.NamedService;
+
+/**
+ * Prototype is generated from prototype blueprint, as is considered the public API.
+ */
+public interface Prototype {
+    /**
+     * Terminating method of this builder that creates a prototype instance.
+     * If the object is a factory, it has a further method {@code build}, that produces the target runtime instance
+     * created from this builder or setup object instance.
+     *
+     * @param <BUILDER>   type of the builder. see {@link #me()}
+     * @param <PROTOTYPE> type of the prototype to be built
+     * @see Prototype.Factory#build()
+     */
+    interface Builder<BUILDER, PROTOTYPE> {
+        /**
+         * Create an instance of the {@link Prototype}. This method is available on
+         * all generated builders for {@link Prototype.Blueprint}.
+         *
+         * @return an instance of the setup object created from this builder.
+         */
+        PROTOTYPE buildPrototype();
+
+        /**
+         * Instance of this builder as the correct type.
+         *
+         * @return this instance typed to correct type
+         */
+        @SuppressWarnings("unchecked")
+        default BUILDER me() {
+            return (BUILDER) this;
+        }
+
+        /**
+         * Resolve an instance to a built instance (if it is a builder).
+         *
+         * @param instance instance to check
+         * @return the same instance if not a builder, or a built instance
+         * @param <T> type of the parameter
+         */
+        @SuppressWarnings("unchecked")
+        default <T> T resolveBuilder(T instance) {
+            if (instance instanceof Builder<?, ?> builder) {
+                return (T) builder.buildPrototype();
+            }
+            if (instance instanceof io.helidon.common.Builder<?, ?> builder) {
+                return (T) builder.build();
+            }
+            return instance;
```

Comment:
that is not the intention of this method. This method just makes sure that if you pass a builder into it, it will build it.
This is to prevent accidentally accepting a builder instance (as it is not validated, and interceptors were not executed).
This has resolved a load of problems in our tests, where we (intentionally) do not call `build` method before passing an instance to another builder.
For example if you call it with a `boolean`, it will just return the boolean as is, but if you call it with `TypeName.Builder`, it will call `build()` on it, to validate the type name is correct.

Reply:
I think we will find there are valid cases where `build()` should not be called before pipelining operations.  I found this to be true during some of my use cases, but I don't recall the details.

Reply:
The build is passing AFAIK, so we do not need it, or I fixed the usages.